usecases/objects/references_add.go

```diff
@@ -22,66 +23,127 @@ import (
 	"github.com/semi-technologies/weaviate/usecases/objects/validation"
 )
 
-// AddObjectReference Class Instance to the connected DB. If the class contains a network
+// AddObjectReference to an existing object. If the class contains a network
 // ref, it has a side-effect on the schema: The schema will be updated to
 // include this particular network ref class.
-func (m *Manager) AddObjectReference(ctx context.Context, principal *models.Principal,
-	id strfmt.UUID, propertyName string, property *models.SingleRef) error {
-	err := m.authorizer.Authorize(principal, "update", fmt.Sprintf("objects/%s", id.String()))
-	if err != nil {
-		return err
+func (m *Manager) AddObjectReference(
+	ctx context.Context,
+	principal *models.Principal,
+	input *AddReferenceInput,
+) *Error {
+	deprecatedEndpoint := input.Class == ""
+	if deprecatedEndpoint { // for backward compatibility only
+		objectRes, err := m.getObjectFromRepo(ctx, "", input.ID, additional.Properties{})
+		if err != nil {
+			errnf := ErrNotFound{} // treated as StatusBadRequest for backward comp
+			if errors.As(err, &errnf) {
+				return &Error{"source object deprecated", StatusBadRequest, err}
+			}
+			return &Error{"source object deprecated", StatusInternalServerError, err}
+		}
+		input.Class = objectRes.Object().Class
+	}
+	path := fmt.Sprintf("objects/%s/%s", input.Class, input.ID)
+	if err := m.authorizer.Authorize(principal, "update", path); err != nil {
+		return &Error{path, StatusForbidden, err}
 	}
 
 	unlock, err := m.locks.LockSchema()
 	if err != nil {
-		return NewErrInternal("could not acquire lock: %v", err)
+		return &Error{"cannot lock", StatusInternalServerError, err}
 	}
 	defer unlock()
 
-	return m.addObjectReferenceToConnectorAndSchema(ctx, principal, id, propertyName, property)
+	validator := validation.New(schema.Schema{}, m.exists, m.config)
```

Comment:
Just a suggestion. For better code reading. I would put the whole logic behind the action of adding a new reference into a separate method so that in `AddObjectReference` there would be the authorizer code and then the invocation of the method that adds a new reference

Reply:
Actually this is a small function and to me it is more readable. The only thing which annoys me is the if block (lines 61-69) which will be removed on later version

Reply:
Cool!